gateway/transport/interceptors/auth/auth.go

```diff
@@ -125,33 +131,109 @@ func (i *interceptor) StreamServerInterceptor(srv any, ss grpc.ServerStream, inf
-		}
-		gwctx := &GatewayContext{
-			UserContext: *userCtx.ToAPIContext(),
-			BearerToken: bearerToken,
-		}
-		gwctx.UserContext.ApiURL = i.idp.ApiURL
-		connectionName := commongrpc.MetaGet(md, "connection-name")
-		conn, err := i.getConnection(connectionName, userCtx)
-		if err != nil {
-			return err
-		}
-		if conn == nil {
-			return status.Errorf(codes.NotFound, "connection not found")
+			if conn == nil {
+				return status.Errorf(codes.NotFound, "connection not found")
+			}
+			gwctx.Connection = *conn
+			ctxVal = gwctx
+		} else {
+			// first we check if the auth method is local, if so, we authenticate the user
+			// using the local auth method, otherwise we use the i.idp.VerifyAccessToken
+			authMethod := appconfig.Get().AuthMethod()
+			var sub string
+			if authMethod == "local" {
+				sessionByToken, err := pglocalauthsession.GetSessionByToken(bearerToken)
+				if err != nil {
+					log.Debugf("failed verifying access token, reason=%v", err)
+					return status.Errorf(codes.Unauthenticated, "invalid authentication")
+				}
+				// TODO change ExpiresAt at the database for date with timezone
+				sessionExpiresAt, err := time.Parse("2006-01-02T15:04:05+00:00", sessionByToken.ExpiresAt)
+
+				if err != nil {
+					log.Debugf("failed verifying access token, reason=%v", err)
+					return status.Errorf(codes.Unauthenticated, "invalid authentication")
+				}
+				if time.Now().After(sessionExpiresAt) {
+					log.Debugf("failed verifying access token, reason=%v", err)
+					return status.Errorf(codes.Unauthenticated, "invalid authentication")
+				}
+
+				user, err := pgusers.GetOneByEmail(sessionByToken.UserEmail)
+				if err != nil {
+					log.Debugf("failed verifying access token, reason=%v", err)
+					return status.Errorf(codes.Unauthenticated, "invalid authentication")
+				}
+				sub = user.Subject
+			} else {
+				sub, err = i.idp.VerifyAccessToken(bearerToken)
+				if err != nil {
+					log.Debugf("failed verifying access token, reason=%v", err)
+					return status.Errorf(codes.Unauthenticated, "invalid authentication")
+				}
+			}
+			if err != nil {
+				log.Debugf("failed verifying access token, reason=%v", err)
+				return status.Errorf(codes.Unauthenticated, "invalid authentication")
+			}
+			userCtx, err := pguserauth.New().FetchUserContext(sub)
+			if userCtx.IsEmpty() {
+				if err != nil {
+					log.Error(err)
+				}
+				return status.Errorf(codes.Unauthenticated, "invalid authentication")
+			}
+			gwctx := &GatewayContext{
+				UserContext: *userCtx.ToAPIContext(),
+				BearerToken: bearerToken,
+			}
+			gwctx.UserContext.ApiURL = i.idp.ApiURL
+			connectionName := commongrpc.MetaGet(md, "connection-name")
+			conn, err := i.getConnection(connectionName, userCtx)
+			if err != nil {
+				return err
+			}
+			if conn == nil {
+				return status.Errorf(codes.NotFound, "connection not found")
+			}
+			gwctx.Connection = *conn
+			ctxVal = gwctx
```

Comment:
This part of the code is repeated, use it once by putting in the outer block of the condition

Reply:
I've removed the else statement, but this is not necessarily a DRY problem. If I move this code to the end of the file, I'd have to move a lot of variables to the top of the block and cause collateral effects on them procedurally; it not only makes the code harder to read but also saves a couple of lines in favor of adding more on the top of the block. Also, the contexts and some other values are slightly different, and making the names different in favor of readability at the cost of some repetition is fairly okay in this case.